### PR TITLE
Improve pageserver/safekeepeer HTTP API errors

### PR DESCRIPTION
**Pull request by sharnoff**
This is part of the work towards making sure that #2419 doesn't happen again.

The work in this PR (at least, so far) is _just_ around making sure that the return codes reflect the nature of the error that occurred. This isn't 100% possible in our current state, because all of our internal functions return an `anyhow::Result`, so in the places where the results are mixed, I've added reminder comments to fix them.

Brief summary of changes:
 * Rename `ApiError::from_error` to `ApiError::from_internal_error`
   * This is the functionality it had before, but the name didn't indicate that.
 * Add `ApiError::from_bad_request`
 * Change `ApiError::BadRequest` and `ApiError::NotFound` to use `anyhow::Error` internally
 * Remove `impl Into<ApiError> for anyhow::Error`
   * Note: this means the try operator (`?`) can no longer be used to implicitly turn an `anyhow::Error` into an `ApiError` in handler functions

There's a lot of subtly changed behavior here. I'm not 100% sure how testing for this should work.

I'd also appreciate some advice on whether it'd be better to make a cleaner (somewhat) PR with new error types supporting the return codes as well, or if this should be merged as soon as there's adequate tests.

### CODE REVIEW

File: libs/utils/src/http/error.rs
Comment (sharnoff) [inline libs/utils/src/http/error.rs:8]: Just noticed - this should be `{0:#?}` instead of `{:?}` so that it doesn't have a bunch of newlines. I'll hold off on changing that for now; want to get general feedback first.

File: libs/utils/src/http/error.rs
Comment (hlinnaka) [inline libs/utils/src/http/error.rs:38]: This is a bit confusing. From the function name and comment, I though that this converts *from* BadRequest *into* `anyhow:Error`.

File: pageserver/src/http/routes.rs
Comment (hlinnaka) [inline pageserver/src/http/routes.rs:212]: This is awkwardly far away from where the `get_tenant` call. Can it be refactored into something more like:
```
let tenant = tenant_mgr::get_tenant(tenant_id, true)
    .map_err(ApiError::NotFound);
```

**Review by hlinnaka — APPROVED** [review]

File: libs/utils/src/http/error.rs
Comment (sharnoff) [reply libs/utils/src/http/error.rs]: Makes sense - I decided to remove both `from_bad_request` and `from_internal_err` so that the way errors are being converted can be a little bit more explicit. To that end, I've also added explicit types to the places where possible tokio task joining errors are handled - for example:

https://github.com/neondatabase/neon/blob/0b53c20685fe82fd35b8cbd5e866acfe0d6161bf/pageserver/src/http/routes.rs#L204-L210

This might be a _bit_ much tbh; I'm not sure where we want the balance of writability vs readability/verboseness to be, but I thought I'd just go for it.

File: pageserver/src/http/routes.rs
Comment (sharnoff) [reply pageserver/src/http/routes.rs]: Ah, good catch. I've applied that to all the existing `map_err` calls

**Comment by sharnoff** [comment]
CI failed only on the same flaky test(s) as before; merging.

**Pull request merged**
